### PR TITLE
rockcraft: init at 1.2.2

### PR DESCRIPTION
**Pull request by jnsgruk**
## Description of changes

This PR introduces the [`rockcraft`](https://canonical-rockcraft.readthedocs-hosted.com/en/latest/) application, which is a tool for building OCI container images.

Before the `rockcraft` package itself is added, there are some more commits adding the required Python libraries. I will follow this PR with more for `snapcraft` and `charmcraft` - hence why I have introduced `_0` and `_1` variants of packages, as those rely on different versions at this time.

I have been maintaining these in my own flake, and they've been stable for some time: https://github.com/jnsgruk/crafts-flake.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by jnsgruk** [comment]
Result of `nixpkgs-review pr 295646` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>24 packages built:</summary>
  <ul>
    <li>python311Packages.craft-application_1</li>
    <li>python311Packages.craft-application_1.dist</li>
    <li>python311Packages.craft-archives</li>
    <li>python311Packages.craft-archives.dist</li>
    <li>python311Packages.craft-cli</li>
    <li>python311Packages.craft-cli.dist</li>
    <li>python311Packages.craft-grammar</li>
    <li>python311Packages.craft-grammar.dist</li>
    <li>python311Packages.craft-parts</li>
    <li>python311Packages.craft-parts.dist</li>
    <li>python311Packages.craft-providers</li>
    <li>python311Packages.craft-providers.dist</li>
    <li>python311Packages.pydantic-yaml_0</li>
    <li>python311Packages.pydantic-yaml_0.dist</li>
    <li>python311Packages.snap-helpers</li>
    <li>python311Packages.snap-helpers.dist</li>
    <li>python311Packages.spdx</li>
    <li>python311Packages.spdx-lookup</li>
    <li>python311Packages.spdx-lookup.dist</li>
    <li>python311Packages.spdx.dist</li>
    <li>python312Packages.spdx</li>
    <li>python312Packages.spdx.dist</li>
    <li>rockcraft</li>
    <li>rockcraft.dist</li>
  </ul>
</details>

File: pkgs/by-name/ro/rockcraft/package.nix [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+python3Packages.buildPythonApplication rec {
+  pname = "rockcraft";
+  version = "1.2.2";
+
+  src = fetchFromGitHub {
+    owner = "canonical";
+    repo = "rockcraft";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-f6qLvk5cebHjZx51nJYYAwILGCTyj6FKlb3HGt6iz4M=";
```
Comment (Lurkki14) [inline pkgs/by-name/ro/rockcraft/package.nix:17]:
```suggestion
    hash = "sha256-f6qLvk5cebHjZx51nJYYAwILGCTyj6FKlb3HGt6iz4M=";
```

File: pkgs/development/python-modules/craft-application_1/default.nix [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  pname = "craft-application_1";
+  version = "1.2.1";
+
+  format = "pyproject";
+
+  src = fetchFromGitHub {
+    owner = "canonical";
+    repo = "craft-application";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-CXZEWVoE66dlQJp4G8tinufjyaDJaH1Muxz/qd/81oA=";
```
Comment (Lurkki14) [inline pkgs/development/python-modules/craft-application_1/default.nix:18]:
```suggestion
    hash = "sha256-CXZEWVoE66dlQJp4G8tinufjyaDJaH1Muxz/qd/81oA=";
```

File: pkgs/development/python-modules/craft-application_1/default.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    craft-cli
+    craft-parts
+    craft-providers
+    pydantic-yaml_0
+    pyyaml
+    setuptools
+    setuptools-scm
+  ];
+
+  env.SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
Comment (Lurkki14) [inline pkgs/development/python-modules/craft-application_1/default.nix:39]:
```suggestion
```
This shouldn't be needed anymore, see https://github.com/NixOS/nixpkgs/commit/8056f9250ce8b7f9bdf0ccdc6c7a6808ae8f1be9

File: pkgs/by-name/ro/rockcraft/package.nix [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  disabledTests = [ "test_expand_extensions" ];
+
+  passthru.updateScript = nix-update-script { };
+
+  meta = {
+    mainProgram = "rockcraft";
+    description = "Create OCI images using the language from Snapcraft and Charmcraft";
+    homepage = "https://github.com/canonical/rockcraft";
+    changelog = "https://github.com/canonical/rockcraft/releases/tag/${version}";
+    license = lib.licenses.lgpl3;
```
Comment (Lurkki14) [inline pkgs/by-name/ro/rockcraft/package.nix:51]: These should be the more specific `lgpl3Plus` or `lgpl3Only`

File: pkgs/development/python-modules/craft-archives/default.nix [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  pname = "craft-archives";
+  version = "1.1.3";
+
+  format = "pyproject";
+
+  src = fetchFromGitHub {
+    owner = "canonical";
+    repo = "craft-archives";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-ZUqMjbOsHwzZyn0NsSTlZTljzagYEirWKEGatXVL43g=";
```
Comment (Lurkki14) [inline pkgs/development/python-modules/craft-archives/default.nix:18]:
```suggestion
    hash = "sha256-ZUqMjbOsHwzZyn0NsSTlZTljzagYEirWKEGatXVL43g=";
```

File: pkgs/development/python-modules/craft-archives/default.nix [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    launchpadlib
+    lazr-restfulclient
+    overrides
+    pydantic_1
+    setuptools
+    setuptools-scm
+    tabulate
+  ];
+
+  env.SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
Comment (Lurkki14) [inline pkgs/development/python-modules/craft-archives/default.nix:39]:
```suggestion
```

File: pkgs/development/python-modules/craft-cli/default.nix [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  pname = "craft-cli";
+  version = "2.5.1";
+
+  format = "pyproject";
+
+  src = fetchFromGitHub {
+    owner = "canonical";
+    repo = "craft-cli";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-yEKF04OPu4paRrghAP78r9hu6cqkUy6z/V7cHNys82I=";
```
Comment (Lurkki14) [inline pkgs/development/python-modules/craft-cli/default.nix:18]:
```suggestion
    hash = "sha256-yEKF04OPu4paRrghAP78r9hu6cqkUy6z/V7cHNys82I=";
```

File: pkgs/development/python-modules/craft-cli/default.nix [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+  propagatedBuildInputs = with python3.pkgs; [
+    platformdirs
+    pydantic_1
+    pyyaml
+    setuptools
+    setuptools-scm
+  ];
+
+  SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
Comment (Lurkki14) [inline pkgs/development/python-modules/craft-cli/default.nix:37]:
```suggestion
```

File: pkgs/development/python-modules/craft-grammar/default.nix [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  pname = "craft-grammar";
+  version = "1.1.2";
+
+  format = "pyproject";
+
+  src = fetchFromGitHub {
+    owner = "canonical";
+    repo = "craft-grammar";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-23KLIO2yHXGe/zb3B8LirJsh+LY9z0c5ZGtF392Kszo";
```
Comment (Lurkki14) [inline pkgs/development/python-modules/craft-grammar/default.nix:18]:
```suggestion
    hash = "sha256-23KLIO2yHXGe/zb3B8LirJsh+LY9z0c5ZGtF392Kszo";
```

File: pkgs/development/python-modules/craft-grammar/default.nix [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  };
+
+  propagatedBuildInputs = with python3.pkgs; [
+    overrides
+    setuptools
+  ];
+
+  env.SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
Comment (Lurkki14) [inline pkgs/development/python-modules/craft-grammar/default.nix:26]:
```suggestion
```

File: pkgs/development/python-modules/craft-parts/default.nix [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  pname = "craft-parts";
+  version = "1.26.2";
+
+  format = "pyproject";
+
+  src = fetchFromGitHub {
+    owner = "canonical";
+    repo = "craft-parts";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-wHv0JWffS916RK4Kgk+FuRthx+ajh0Ka4DBwGrLdUBs=";
```
Comment (Lurkki14) [inline pkgs/development/python-modules/craft-parts/default.nix:18]:
```suggestion
    hash = "sha256-wHv0JWffS916RK4Kgk+FuRthx+ajh0Ka4DBwGrLdUBs=";
```

File: pkgs/development/python-modules/craft-providers/default.nix [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+  pname = "craft-providers";
+  version = "1.23.0";
+
+  format = "pyproject";
+
+  src = fetchFromGitHub {
+    owner = "canonical";
+    repo = "craft-providers";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-9ZoNgpuGytwozRsw0wnS3d2UBOIsh3VI/uzB1RD2Zac=";
```
Comment (Lurkki14) [inline pkgs/development/python-modules/craft-providers/default.nix:18]:
```suggestion
    hash = "sha256-9ZoNgpuGytwozRsw0wnS3d2UBOIsh3VI/uzB1RD2Zac=";
```

File: pkgs/development/python-modules/craft-providers/default.nix [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    platformdirs
+    pydantic_1
+    pyyaml
+    requests-unixsocket
+    setuptools
+    setuptools-scm
+    urllib3
+  ];
+
+  SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
Comment (Lurkki14) [inline pkgs/development/python-modules/craft-providers/default.nix:48]:
```suggestion
```

File: pkgs/development/python-modules/pydantic-yaml_0/default.nix [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+  pname = "pydantic_yaml_0";
+  version = "0.11.2";
+
+  format = "pyproject";
+
+  src = fetchFromGitHub {
+    owner = "NowanIlfideme";
+    repo = "pydantic-yaml";
+    rev = "refs/tags/v${version}";
+    sha256 = "sha256-AeUyVav0/k4Fz69Qizn4hcJKoi/CDR9eUan/nJhWsDY=";
```
Comment (Lurkki14) [inline pkgs/development/python-modules/pydantic-yaml_0/default.nix:18]:
```suggestion
    hash = "sha256-AeUyVav0/k4Fz69Qizn4hcJKoi/CDR9eUan/nJhWsDY=";
```

File: pkgs/development/python-modules/pydantic-yaml_0/default.nix [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    importlib-metadata
+    pydantic_1
+    ruamel-yaml
+    semver
+    types-deprecated
+    setuptools
+    setuptools-scm
+  ];
+
+  env.SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
Comment (Lurkki14) [inline pkgs/development/python-modules/pydantic-yaml_0/default.nix:40]:
```suggestion
```

File: pkgs/development/python-modules/snap-helpers/default.nix [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+  pname = "snap-helpers";
+  version = "0.4.2";
+
+  format = "pyproject";
+
+  src = fetchFromGitHub {
+    owner = "albertodonato";
+    repo = "snap-helpers";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-7JBvrD4WNOcFSVx3xauk4JgiVGKWuUEdUMRlH7mudE4=";
```
Comment (Lurkki14) [inline pkgs/development/python-modules/snap-helpers/default.nix:18]:
```suggestion
    hash = "sha256-7JBvrD4WNOcFSVx3xauk4JgiVGKWuUEdUMRlH7mudE4=";
```

File: pkgs/development/python-modules/snap-helpers/default.nix [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+  };
+
+  propagatedBuildInputs = with python3.pkgs; [
+    packaging
+    pyyaml
+    setuptools
+  ];
+
+  SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
Comment (Lurkki14) [inline pkgs/development/python-modules/snap-helpers/default.nix:27]:
```suggestion
```

**Review by Lurkki14 — CHANGES_REQUESTED** [review]

**Comment by jnsgruk** [comment]
@Lurkki14 nice, thank you! I've made those changes and it all still seems to build just fine! Thanks for the tip on the new setuptools-scm hook :slightly_smiling_face:

**Comment by jnsgruk** [comment]
@RaitoBezarius @SuperSandro2000 could I get a little help here please? :slightly_smiling_face:

File: pkgs/development/python-modules/spdx-lookup/default.nix [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+  version = "0.3.2";
+
+  src = fetchFromGitHub {
+    owner = "bbqsrc";
+    repo = "spdx-lookup-python";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-jtzhqRAj1BWdU8AuR7Gr343mL5alLXhi+SyCkCI5AAU=";
+  };
+
+  propagatedBuildInputs = with python3.pkgs; [
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/spdx-lookup/default.nix:19]: We cannot use python3 inside of pythonPackages. Please use the packages directly as inputs

File: pkgs/development/python-modules/snap-helpers/default.nix [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+  format = "pyproject";
+
+  src = fetchFromGitHub {
+    owner = "albertodonato";
+    repo = "snap-helpers";
+    rev = "refs/tags/${version}";
+    hash = "sha256-7JBvrD4WNOcFSVx3xauk4JgiVGKWuUEdUMRlH7mudE4=";
+  };
+
+  propagatedBuildInputs = with python3.pkgs; [
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/snap-helpers/default.nix:21]: same here

File: pkgs/development/python-modules/snap-helpers/default.nix [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+, fetchFromGitHub
+, nix-update-script
+, python3
+}:
+
+buildPythonPackage rec {
+  pname = "snap-helpers";
+  version = "0.4.2";
+
+  format = "pyproject";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/snap-helpers/default.nix:12]: This should be ``pyproject = true``

File: pkgs/development/python-modules/snap-helpers/default.nix [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+    rev = "refs/tags/${version}";
+    hash = "sha256-7JBvrD4WNOcFSVx3xauk4JgiVGKWuUEdUMRlH7mudE4=";
+  };
+
+  propagatedBuildInputs = with python3.pkgs; [
+    packaging
+    pyyaml
+    setuptools
+  ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/snap-helpers/default.nix:26]: Please execute tests https://github.com/albertodonato/snap-helpers/tree/main/tests

File: pkgs/development/python-modules/snap-helpers/default.nix [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+  src = fetchFromGitHub {
+    owner = "albertodonato";
+    repo = "snap-helpers";
+    rev = "refs/tags/${version}";
+    hash = "sha256-7JBvrD4WNOcFSVx3xauk4JgiVGKWuUEdUMRlH7mudE4=";
+  };
+
+  propagatedBuildInputs = with python3.pkgs; [
+    packaging
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/snap-helpers/default.nix:22]: Not sure if packaging is required at all

File: pkgs/development/python-modules/craft-providers/default.nix [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+
+  postPatch = ''
+    substituteInPlace craft_providers/__init__.py \
+      --replace-fail "dev" "${version}"
+
+    # The urllib3 incompat: https://github.com/msabramo/requests-unixsocket/pull/69
+    # This is already patched in nixpkgs.
+    substituteInPlace pyproject.toml \
+      --replace-fail "setuptools==67.8.0" "setuptools" \
+      --replace-fail "urllib3<2" "urllib3"
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/craft-providers/default.nix:35]: Can we use relaxDepsHook here?

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please add pythonImportsCheck to all packages. Also if there are tests, we should execute them. I didn't check all packages for them.

I also stopped commenting the same thing on the second occurrence or so.

File: pkgs/development/python-modules/craft-providers/default.nix (hunk 23 above)
Comment (jnsgruk) [reply pkgs/development/python-modules/craft-providers/default.nix:35]: I can't seem to make that work with pyproject.toml projects - tried a bunch of different configs with it in this project and couldn't make it happen.

**Comment by jnsgruk** [comment]
@SuperSandro2000 thank you! I've addressed your comments all bar one on `pythonRelaxDepsHook` which didn't seem to work for these cases.

File: pkgs/by-name/ro/rockcraft/package.nix [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+  preCheck = ''
+    mkdir -p check-phase
+    export HOME="$(pwd)/check-phase"
+  '';
+
+  disabledTests = [ "test_expand_extensions" ];
+
+  passthru.updateScript = nix-update-script { };
+
+  meta = {
```
Comment (drupol) [inline pkgs/by-name/ro/rockcraft/package.nix:49]: Missing `meta.mainProgram`.

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ro/rockcraft/package.nix (hunk 24 above)
Comment (drupol) [reply pkgs/by-name/ro/rockcraft/package.nix:49]: /me needs some sleep.

**Review by drupol — APPROVED** [review]

**Review by drupol — APPROVED** [review]
LGTM now! thanks for making the changes.

File: pkgs/development/python-modules/craft-providers/default.nix (hunk 23 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/craft-providers/default.nix:50]: That's fine, we don't need to use it.

File: pkgs/development/python-modules/craft-parts/default.nix [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+  };
+
+  patches = [
+    ./bash-path.patch
+  ];
+
+  postPatch = ''
+    # substituteInPlace craft_parts/__init__.py \
+    #   --replace-fail "dev" "${version}"
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/craft-parts/default.nix:46]: Do we still need that?

**Comment by jnsgruk** [comment]
@SuperSandro2000 can I beg another review please? Thanks in advance 🙏

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by jnsgruk** [comment]
Thank you! 🤩

File: pkgs/top-level/python-packages.nix [hunk 26]
```diff
@@ -10529,6 +10541,8 @@ self: super: with self; {
 
   pydantic-settings = callPackage ../development/python-modules/pydantic-settings { };
 
+  pydantic-yaml-0 = callPackage ../development/python-modules/pydantic-yaml-0 { };
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix:10544]: We only allow multiple versions in `python-packages.nix` in very rare cases. Why was this done?
Anyway, it should have been
```suggestion
pydantic-yaml_0 = callPackage ../development/python-modules/pydantic-yaml/0.nix { };
```

File: pkgs/development/python-modules/pydantic-yaml_0/default.nix [hunk 27]
```diff
@@ -0,0 +1,50 @@
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+, nix-update-script
+, python3
+}:
+
+buildPythonPackage rec {
+  pname = "pydantic_yaml_0";
```
[the remark this answers is not in the document]
Comment (dotlambda) [reply pkgs/development/python-modules/pydantic-yaml_0/default.nix]: But https://pypi.org/project/pydantic-yaml-0/ doesn't exist.

File: pkgs/top-level/python-packages.nix (hunk 26 above)
Comment (dotlambda) [reply pkgs/top-level/python-packages.nix:10544]: I'm rectifying this in https://github.com/NixOS/nixpkgs/pull/329667.

Comment (jnsgruk) [reply pkgs/top-level/python-packages.nix:10544]: Hey! Nice, thank you. I'll take a look!

**Pull request merged**